modules/autohalfop.cpp

```diff
@@ -0,0 +1,646 @@
+        MCString::iterator itQueue = m_msQueue.find(Nick.GetNick().AsLower());
+
+        if (itQueue == m_msQueue.end()) {
+            PutModule(
+                t_f("[{1}] sent an unchallenged response.  This could be due "
+                    "to lag.")(Nick.GetHostMask()));
+            return false;
+        }
+
+        CString sChallenge = itQueue->second;
+        m_msQueue.erase(itQueue);
+
+        for (const auto& it : m_msUsers) {
+            if (it.second->HostMatches(Nick.GetHostMask())) {
+                if (sResponse ==
+                    CString(it.second->GetUserKey() + "::" + sChallenge)
+                        .MD5()) {
+                    HalfOpUser(Nick, *it.second);
+                    return true;
+                } else {
+                    PutModule(
+                        t_f("WARNING! [{1}] sent a bad response.  Please "
+                            "verify that you have their correct password.")(
+                            Nick.GetHostMask()));
+                    return false;
+                }
+            }
+        }
+
+        PutModule(
+            t_f("WARNING! [{1}] sent a response but did not match any defined "
+                "users.")(Nick.GetHostMask()));
+        return false;
+    }
+
+    void ProcessQueue() {
+        bool bRemoved = true;
+
+        // First remove any stale challenges
+
+        while (bRemoved) {
+            bRemoved = false;
+
+            for (MCString::iterator it = m_msQueue.begin();
+                 it != m_msQueue.end(); ++it) {
+                if (!it->second.empty()) {
+                    m_msQueue.erase(it);
+                    bRemoved = true;
+                    break;
+                }
+            }
+        }
+
+        // Now issue challenges for the new users in the queue
+        for (auto& it : m_msQueue) {
+            it.second = CString::RandomString(AUTOHALFOP_CHALLENGE_LENGTH);
+            PutIRC("NOTICE " + it.first + " :!ZNCAHO CHALLENGE " + it.second);
+        }
+    }
+
+    void HalfOpUser(const CNick& Nick, const CAutoHalfOpUser& User) {
+        const vector<CChan*>& Chans = GetNetwork()->GetChans();
+
+        for (CChan* pChan : Chans) {
+            if (pChan->HasPerm(CChan::HalfOp) &&
+                User.ChannelMatches(pChan->GetName())) {
+                const CNick* pNick = pChan->FindNick(Nick.GetNick());
+
+                if (pNick && !pNick->HasPerm(CChan::HalfOp)) {
+                    PutIRC("MODE " + pChan->GetName() + " +h " +
+                           Nick.GetNick());
+                }
+            }
+        }
+    }
+
+  private:
+    map<CString, CAutoHalfOpUser*> m_msUsers;
+    MCString m_msQueue;
+};
+
+void CAutoHalfOpTimer::RunJob() { m_pParent->ProcessQueue(); }
+
+template <>
+void TModInfo<CAutoHalfOpMod>(CModInfo& Info) {
+    Info.SetWikiPage("autohalfop");
+}
+
+NETWORKMODULEDEFS(CAutoHalfOpMod, t_s("Auto halfop the good people"))
```

Comment:
We shouldn't copypaste/duplicate code like that. A better way would be to have a single module which supports ops, voices, halfops

Reply:
Perhaps that shall be my second contribution.

Reply:
Perhaps it shall.

I believe merging this PR can wait until that's done though.